JSDoc automation

garbados opened the pull request:
<!--
Thank you for submitting a Pull Request.

This code is leveraged to monitor critical services. Please consider the following:
* Tests are required.
* Performance matters.
* Features that are specific to just your app are unlikely to make it in.

Please fill out the relevant sections as follows:
* Proposed Release Notes: Bulleted list of recommended release notes for the change(s).
* Links: Any relevant links for the change.
* Details: In-depth description of changes, other technical notes, etc.
-->

## Proposed Release Notes

## Links

- Closes https://github.com/newrelic/node-newrelic/issues/593

## Details

This PR adds a CI job for publishing current documentation to GH Pages during release. I have done this by adding a job to `.github/workflows/publish-branch.yml`. Because this integrates with a CI, I have only been able to subject it to limited testing using [act](https://github.com/nektos/act).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/newrelic/node-newrelic?pullRequest=908) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/newrelic/node-newrelic?pullRequest=908) before we can accept your contribution.<br/>**5** out of **7** committers have signed the CLA.<br/><br/>:white_check_mark: lykkin<br/>:white_check_mark: martinkuba<br/>:white_check_mark: garbados<br/>:white_check_mark: nijotz<br/>:white_check_mark: michaelgoin<br/>:x: NatalieWolfe<br/>:x: Peter Svetlichny<br/><hr/>**Peter Svetlichny** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/newrelic/node-newrelic?pullRequest=908) it.</sub>

garbados commented:
Got some weird commit history shenanigans going on. Hm...

michaelgoin commented on .github/workflows/publish-branch.yml:
```diff
@@ -23,3 +23,9 @@ jobs:
     - run: npm publish
       env:
         NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
+    - run: npm run public-docs
+    - name: Deploy Docs
+      uses: peaceiris/actions-gh-pages@v3
+      with:
+        github_token: ${{ secrets.GITHUB_TOKEN }}
+        publish_dir: ./out
```
This is not actually the workflow you want. This is sort of a fall-back quick publish as needed workflow. Which maybe we don't really need as we can always manually publish. Anyways, this is not ran as the part of a standard release.

The create-release workflow is the one you want. It creates a tag, publishes to npm and then creates a GitHub release. I'd prob add this after the release creation in the workflow.

garbados replied:
Ah, great. Thanks for the clarification. I'll move the job around.

bizob2828 commented on .eslintrc.js:
```diff
@@ -17,5 +17,6 @@ module.exports = {
     }
-  ]
+  ],
+  plugins: ['jsdoc']
```
this def introduces a ton of warnings but personally I think it's ok to address these as files are touched, thoughts @michaelgoin ?

bizob2828 commented on .github/workflows/create-release.yml:
```diff
@@ -39,3 +39,9 @@ jobs:
       run: node ./bin/create-github-release.js --tag ${{ steps.get_tag.outputs.latest_tag }} --repo-owner ${{ github.repository_owner }}
       env:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    - run: npm run public-docs
```
how did you test this or did you just trust the action will just work ™️

garbados replied:
Just using [act](https://github.com/nektos/act). Naturally this can only test publishing actions as dry runs, leading to a certain unfortunate distance between testing and production.

bizob2828 reviewed (commented):
just wanted to throw my 2 cents regarding the new eslint plugin

garbados commented on .eslintrc.js:
```diff
@@ -6,7 +6,7 @@
 
 module.exports = {
-  extends: '@newrelic',
+  extends: ['@newrelic', 'plugin:jsdoc/recommended'],
```
Will do.

bizob2828 commented on .eslintrc.js:
```diff
@@ -16,6 +18,12 @@ module.exports = {
       }
+    },
+    {
+      files: ['test/**/*.js'],
```
we have a .eslintrc.js in test. it might be more appropriate to put this in that config

garbados replied:
Fixed.

garbados commented:
OK @bizob2828 this should be ready for review.

bizob2828 reviewed (changes requested):
I apologize for the back and forth on this one.  I just want to make sure we go out the gate with the jsdoc linting on the right foot, otherwise it'll get ignored.

michaelgoin commented on .eslintrc.js:
```diff
@@ -7,8 +7,7 @@
 
 module.exports = {
   extends: ['@newrelic', 'plugin:jsdoc/recommended'],
```
If you look in package.json at the "public-docs" script you'll see what we publish.

So ideally we wouldn't require jsdoc on files that aren't those ones / in those locations. But I do think we are fine wanting all jsdoc to be valid wherever it lives.

garbados replied:
Okiedoke, so extending the lint task to include public facing docs raises some linting errors. I can resolve them in a two-commit process so we get the lint script changes separately from the code changes to satisfy those rules.

bizob2828 replied:
ok. i think one thing you forgot to do is to disable `'jsdoc/require-jsdoc': 'off'` at the base and only require jsdoc for the public facing things.

bizob2828 replied:
it's prob best to file a ticket to add defined types to the eslint rules so we don't get `jsdoc/no-undefined-types` warnings

garbados commented:
Hey, no sweat about the back and forth. That's just how it goes :)

I've updated the eslint task to include the examples folder, but I've also added an `examples/.eslintrc.js` file to manage the ESLint rules we want for our examples. This is mostly identical to the top-level config, but turns off no-console and node/no-extraneous-require because examples sometimes include console statements and that's ok, and `require('newrelic')` is pretty standard in our examples but trips that extraneous require rule.

To fix the ESLint errors that came up during linting, I made some changes to our examples that included additional comments and a little restructuring here and there for clarity's sake. I verified that these new implementations worked in a `node-newrelic-examples` test dir that I have set up locally, which includes a `newrelic.js` config file that lets me run examples as `node -r newrelic -e "\`pbpaste\`"` or `node -r newrelic $FILENAME`.

Let me know if there's anything else I should do on this :)

bizob2828 commented on examples/api/background-transactions/example1-basic.js:
```diff
@@ -24,13 +24,16 @@ var transactionName = 'myCustomTransaction'
 
 // Function to simulate async work.
+/**
+ * @param callback
```
this is unfortunate that this is valid jsdoc.  knowing the type is what i really find useful

garbados replied:
Types added :)

bizob2828 commented on lib/collector/remote-method.js:
```diff
@@ -29,6 +29,11 @@ const USER_AGENT_FORMAT = 'NewRelic-NodeAgent/%s (nodejs %s %s-%s)'
 const ENCODING_HEADER = 'CONTENT-ENCODING'
 const DEFAULT_ENCODING = 'identity'
 
+/**
```
did you just start addressing warnings? I don't see these as public facing but maybe I'm not understanding the `public-docs` npm script

garbados replied:
I think these were automatically added by mistake, possibly by husky. My apologies -- I'll revert these changes.

bizob2828 replied:
that's fine.  if husky does it then we can't avoid it. I'm wondering which rule is auto fixing to an empty multiline comment stanza

garbados replied:
Changes to `lib/` reverted :)

bizob2828 commented on lib/collector/remote-method.js:
```diff
@@ -142,6 +147,9 @@ RemoteMethod.prototype._post = function _post(data, nrHeaders, callback) {
     makeRequest()
   }
 
+  /**
```
does this remove warnings, if so that kinda stinks

garbados replied:
This specifically has been removed but also these "empty comment blocks" have been made non-empty in all our public-facing docs.

bizob2828 reviewed (commented):
overall looks good. we're getting closer for sure.  lmk if you need help addressing or unclear about some comments

bizob2828 commented on .eslintrc.js:
```diff
@@ -6,16 +6,24 @@
       rules: {
         'header/header': ['off']
       }
+    },
```
@garbados there was def some miscommunication happening via this PR comments.  I just went ahead and addressed what I was trying to explain.  Let's go over this when you are online today

bizob2828 commented on test/.eslintrc.js:
```diff
@@ -5,6 +5,11 @@
 
 'use strict'
 module.exports = {
+  plugins: ['disable'],
```
I pared this down because .eslintrc inherit from parents so you really only have to load or override what is different

garbados replied:
Thanks, I appreciate it!

bizob2828 reviewed (approved):
This is great.  Thanks for taking the time to wire all of this up.  Can you make sure to log the ticket to add the definedTypes in .eslintrc for our first class "types"?

michaelgoin commented on .eslintrc.js:
```diff
@@ -6,16 +6,24 @@
+    {
+      files: ['./lib/shim/*.js', 'lib/transaction/handle.js', 'api.js'],
+      rules: {
+        'jsdoc/require-jsdoc': 'warn'
```
Wondering if this should be 'error'.  Easy enough to change later.

garbados replied:
Switching this to 'error' highlights 40 errors. Is that acceptable? Should I resolve them as part of this PR?

michaelgoin replied:
Ah, given all the examples ended up with jsdoc I assumed this was covered. Lets land as warn then and then follow-up with switching to error and fixing things.

garbados replied:
Can do 👍

michaelgoin commented on .github/workflows/create-release.yml:
```diff
@@ -39,3 +39,9 @@ jobs:
       run: node ./bin/create-github-release.js --tag ${{ steps.get_tag.outputs.latest_tag }} --repo-owner ${{ github.repository_owner }}
       env:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    - run: npm run public-docs
+    - name: Deploy Docs
+      uses: peaceiris/actions-gh-pages@v3
+      with:
+        github_token: ${{ secrets.GITHUB_TOKEN }}
+        publish_dir: ./out
```
Curious why we aren't using `npm run publish-docs`?

Could be it doesn't work, just want to understand.

My only concern being using CI specific mechanisms can make things harder in the long run when you occasionally have to do things manually, need to debug, etc. Ideal state to me is to keep local/CI runs as close together as possible and leverage specific CI platform just for the specific glue needed. Not always possible, but a guidepost.

garbados replied:
Valid, I'll make the change.

garbados replied:
My only concern is: do we know that the CI box will have appropriate credentials? Do we need to make any changes so that it does?

michaelgoin replied:
Probably something that could be tested out on a fork like I did with the prepare release script.

That being said, the prepare release workflow is a good example of how to setup the credentials needed for pushing code and whatnot: https://github.com/newrelic/node-newrelic/blob/main/.github/workflows/prepare-release.yml#L28-L35

It sets up the GitHub user for doing commits and ensures the token is set in the ENV for the API calls. I'm guessing this only needs the creds setup but don't know for sure w/o testing.

michaelgoin commented on examples/api/segments/example1-callbacks.js:
```diff
@@ -5,12 +5,53 @@
+  })
+}
+
+// then we stub out the task that handles that task's result,
+// to show how the result is passed throughthe segment handler
+/**
+ * @param {string} greetings
+ * @param {Function} callback
+ */
```
I feel like the jsdoc on all these example files is just getting in the way of showing stuff to the user but we can land and then later remove if needed.

garbados replied:
Yeah, I understand that perspective. I could take it or leave it. It's sort of unusual to have JSDoc comments in examples but more clarity is probably better than less, generally speaking.

michaelgoin replied:
More clarity is good when theres clarity being added. Calling out that "greetings" and "callback" are parameters for generically meeting jsdoc's doesn't really add any clarity. I can already read that in the function. I think they are just noise getting in the way of understanding what we are trying to communicate.

That being said, I don't feel strongly enough that it needs to be handled via this PR.

garbados replied:
No problem. I think you're right on the money. It's no trouble to replace these blocks with `eslint-disable-next-line` 👍

garbados commented on .lintstagedrc.json:
```diff
@@ -1,5 +1,7 @@
 {
-  "**/*.js": [
-    "eslint --fix"
```
Just for anyone who's curious, I made this change because otherwise the linter would get upset at the JS bundled into the docs site.

michaelgoin replied:
what i might recommend instead is updating `.eslintignore` to add `docs/scripts/*`. then we don't have to bend over backwards to ignore things.

what you have works, and there are a few other styles possible, but reading through their documentation it seems their opinion is to push ignoring of files into the specific tooling you are using.

https://www.npmjs.com/package/lint-staged#ignoring-files

> and ignoring any files should be configured in the task itself.

michaelgoin commented on examples/.eslintrc.js:
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright 2020 New Relic Corporation. All rights reserved.
```
```suggestion
 * Copyright 2021 New Relic Corporation. All rights reserved.
```

michaelgoin commented on bin/publish-docs.sh:
```diff
@@ -5,12 +5,12 @@
 git commit -m "docs: update for ${PACKAGE_VERSION}"
-git push origin gh-pages && git push public gh-pages:gh-pages
\ No newline at end of file
+git push origin gh-pages
```
👍 

The original script assumed you were running locally on a repo who's origin pointed to GitHub Enterprise and had a remote of 'public' pointing to the public repo mirror. Now we are just working off of public GitHub and are assuming running on a clean pull of the repo.
